src/cargo/sources/registry/remote.rs

```diff
@@ -13,16 +15,23 @@ use ops;
 use sources::git;
 use sources::registry::{RegistryData, RegistryConfig, INDEX_LOCK};
 use util::network;
-use util::paths;
-use util::{FileLock, Filesystem};
+use util::{FileLock, Filesystem, LazyCell};
 use util::{Config, CargoResult, ChainError, human, Sha256, ToUrl};
 
 pub struct RemoteRegistry<'cfg> {
     index_path: Filesystem,
     cache_path: Filesystem,
     source_id: SourceId,
     config: &'cfg Config,
-    handle: Option<Easy>,
+    handle: LazyCell<RefCell<Easy>>,
+
+    // Note that this `'static lifetime here is actually a lie, it's actually a
+    // borrow into the `repo` object below. We're guaranteed, though, that if
+    // filled in `tree` will be destroyed first, so this should be ok.
+    tree: LazyCell<RefCell<Option<git2::Tree<'static>>>>,
```

Comment:
Why do we need `LazyCell` on top of `RefCell`? `LazyCell` is useful to return `&T` and not `Ref<T>`, but we are returning a `Ref` anyway, so just `RefCell<Option<git2::Tree<'static>>>` should be enough...